### ISSUE TITLE
Output of RandomGhosting is not float32

### ISSUE THREAD

**Issue by Linardos** [fepegar/torchio#367]
As I am still trying to wrap my head around the SubjectsDataset class I find myself confused at this error I've been getting when I try to do training with a defined batch_size on the Dataloader (using torch.utils.data.Dataloader):

`RuntimeError: Expected object of scalar type Float but got scalar type Double for sequence element 1 in sequence argument at position #1 'tensors'`

If I don't set a batch_size this error disappears. I found the exact same issue in PyTorch forums and the solution involves making a transformation ToTensor that explicitly defines the data type to be loaded. 
https://discuss.pytorch.org/t/runtimeerror-expected-object-of-scalar-type-double-but-got-scalar-type-long-for-sequence-element-1-in-sequence-argument-at-position-1-tensors/46952

I thought I could do this using the Lambda transformation; however, it's not clear to me how as the images are ScalarImage type.
for example this is not possible: 
`subject['gt'].tensor = subject['gt'].tensor.type(torch.DoubleTensor) `

Thoughts?

**Comment by fepegar** [comment]
Hi, @Linardos. It's hard to understand what's going on with so little information and without a specific question. Can you post the full error using the `python-traceback` syntax highlighting? It would be great if you could share a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) so I can try to get the error on my own computer.

Anyway, this looks more like a PyTorch issue. Are you missing a conversion to float before feeding a tensor to your model? I.e. `tensor = tensor.float()`.

**Comment by Linardos** [comment]
Ok it was a good idea to make this minimal example indeed. After I started trimming stuff I identified the source of my issue as the RandomGhosting transformation. 

It's not really a PyTorch issue because it happens before I load into the model, it happens as soon as I iterate over the batches. Here's my minimal example. If you can make a toy data to replace mine you can have a look for yourself. As soon as I commented out RandomGhosting, the error disappears
`class MNMDataLoader():
    def __init__(self, data_owner=None, images_dir=TRAIN_FOLDER):
        # get all subjects from one folder
        print(f"Initializing Data Loader for {data_owner}") if data_owner is not None else print("Initialized Data Loader")
        print(f"Loading {images_dir}")
        info = pd.read_csv(INFO_FILE)
        subject_names = sorted([child.name for child in Path.iterdir(images_dir)
                                if Path.is_dir(child)])

        self.subjects = []
        self.pathologies = []
        for subject in subject_names:
            case = info.loc[info[CODE] == subject]
            # load this image
            case_code = case[CODE].item()
            if case[PATHOLOGY].item() not in config['data']['pathologies']:
                continue
            mris = [images_dir / case_code / str(case_code + suffix) for suffix in [ED_SUFFIX, ES_SUFFIX]]
            gts = [images_dir / case_code / str(case_code + suffix) for suffix in [ED_GT_SUFFIX, ES_GT_SUFFIX]]
            for mri, gt in zip(mris, gts):
                subject = tio.Subject(
                    mri=tio.ScalarImage(mri),
                    gt=tio.LabelMap(gt),
                    )
                self.subjects.append(subject)
            self.pathologies.append(case[PATHOLOGY].item())

    
    def __len__(self):
        return len(self.subjects)

    def __getitem__(self, index):
        return self.subjects[index], self.pathologies[index]
    
    def get_dataset(self):
        # Transform and define dataset object

        training_transform = Compose([
            CropOrPad((150,150,10), mask_name='gt'),
            OneOf([
                RandomFlip(axes=('LR', 'SI')),
                RandomAffine(translation=0),
                # intensity
                RandomMotion(),
                RandomGhosting(), # THIS IS THE ISSUE
                RandomSpike(),
                RandomBiasField(),
                RandomBlur(),
                RandomNoise(),
                RandomSwap(patch_size=(16, 16, 1)),
    #             RandomGamma()
            ], p=0.5)


        ])

        validation_transform = Compose([
            CropOrPad((150,150,10), mask_name='gt')
        ])

        training_split_ratio = 0.9
        num_subjects = self.__len__()
        print('Dataset size:', num_subjects, 'images')
        num_training_subjects = int(training_split_ratio * num_subjects)

        training_subjects = self.subjects[:num_training_subjects]
        validation_subjects = self.subjects[num_training_subjects:]
        
        training_set = tio.SubjectsDataset(
            training_subjects, transform=training_transform)

        validation_set = tio.SubjectsDataset(
            validation_subjects, transform=validation_transform)

        print('Training set:', len(training_set), 'subjects')
        print('Validation set:', len(validation_set), 'subjects')
        return training_set, validation_set 

    def load(self, b_size):
        training_set, validation_set = self.get_dataset()
        training_loader = torch.utils.data.DataLoader(
            training_set, batch_size=b_size)

        validation_loader = torch.utils.data.DataLoader(
            validation_set, batch_size=b_size)
        return training_loader, validation_loader

dl = MNMDataLoader()
tr, va = dl.load(b_size=1)
for batch in tr:
    print("this works")
    break

dl = MNMDataLoader()
tr, va = dl.load(b_size=20)
for batch in tr:
    print("this doesn't")`

**Comment by fepegar** [comment]
Thanks for sharing your code. Note this is not a minimal working example, because it would not work if I ran it on my computer!

But your investigation about RandomGhosting is very helpful. I can reproduce that:

```python
In [1]: import torchio as tio                                                                                                                                                                                                                                         

In [2]: colin = tio.datasets.Colin27()                                                                                                                                                                                                                                

In [3]: transformed = tio.RandomGhosting()(colin)                                                                                                                                                                                                                     

In [4]: transformed.t1.data.dtype                                                                                                                                                                                                                                     
Out[4]: torch.float64
```

I'll fix that in a minute.

**Comment by Linardos** [comment]
Yeah I guess it wasn't that minimal, sorry hahah. But since we found the issue it's fine. Thanks for your quick replies.

**Comment by fepegar** [comment]
Fixed in `v0.18.1`. Thanks for reporting!
